From 2a6cf5220c171d848903902a1d94979849259400 Mon Sep 17 00:00:00 2001
From: Stephane Eranian <eranian@google.com>
Date: Mon, 23 Apr 2012 18:03:28 +0200
Subject: [PATCH 1/2] perf tools: add support for libpfm4

This patch links perf with the libpfm4 library.
This library contains all the hardware event tables for all
processors supported by perf_events. This is a helper library
that help convert from a symbolic event name to the event
encoding required by the underlying kernel interface. This
library is open-source and available from: http://perfmon2.sf.net.

With this patch, it is possible to specify full hardware event
by names. Hardware filters are also supported. Events must be
specified via the --pfm-events and not -e option. Both options
are active at the same time and it is possible to mix and match:

$ perf stat --pfm-events inst_retired:any_p:c=1:i -e cycles ....

You need to have libpfm4-dev package installed on your system.

Signed-off-by: Stephane Eranian <eranian@google.com>

Conflicts:

	tools/perf/builtin-record.c
	tools/perf/builtin-stat.c
---
 tools/perf/Documentation/perf-record.txt |    9 ++
 tools/perf/Documentation/perf-stat.txt   |    8 ++
 tools/perf/Makefile                      |   10 ++
 tools/perf/builtin-list.c                |   12 ++
 tools/perf/builtin-record.c              |   17 +++
 tools/perf/builtin-stat.c                |   17 +++
 tools/perf/config/feature-tests.mak      |   15 ++-
 tools/perf/util/parse-events.c           |  177 ++++++++++++++++++++++++++++++
 tools/perf/util/parse-events.h           |    5 +
 9 files changed, 269 insertions(+), 1 deletion(-)

diff --git a/tools/perf/Documentation/perf-record.txt b/tools/perf/Documentation/perf-record.txt
index b38a1f9..23d060e 100644
--- a/tools/perf/Documentation/perf-record.txt
+++ b/tools/perf/Documentation/perf-record.txt
@@ -182,6 +182,15 @@ is enabled for all the sampling events. The sampled branch type is the same for
 The various filters must be specified as a comma separated list: --branch-filter any_ret,u,k
 Note that this feature may not be available on all processors.
 
+--pfm-events event::
+this option is only available when perf is linked with the libpfm4 library
+(see http://perfmon2.sf.net). It allows passing hardware events as strings
+for all support processors. Event filters can also be used. As an example:
+perf record --pfm-events inst_retired:any_p:u:c=1:i. More than one event can
+be passed to the option using the comma separator. Hardware events and
+generic hardware events cannot be mixed together. The latter must be used
+with the -e option. The -e option and this one can be mixed and matched.
+
 SEE ALSO
 --------
 linkperf:perf-stat[1], linkperf:perf-list[1]
diff --git a/tools/perf/Documentation/perf-stat.txt b/tools/perf/Documentation/perf-stat.txt
index 2fa173b..29b1d12 100644
--- a/tools/perf/Documentation/perf-stat.txt
+++ b/tools/perf/Documentation/perf-stat.txt
@@ -109,6 +109,14 @@ Log output to fd, instead of stderr.  Complementary to --output, and mutually ex
      3>>results perf stat --log-fd 3 --append -- $cmd
 
 
+--pfm-events events::
+this option is only available when perf is linked with the libpfm4 library
+(see http://perfmon2.sf.net). It allows passing hardware events as strings
+for all support processors. Event filters can also be used. As an example:
+perf stat --pfm-events inst_retired:any_p:u:c=1:i. More than one event can
+be passed to the option using the comma separator. Hardware events and
+generic hardware events cannot be mixed together. The latter must be used
+with the -e option. The -e option and this one can be mixed and matched.
 
 EXAMPLES
 --------
diff --git a/tools/perf/Makefile b/tools/perf/Makefile
index 0a619af..7f36975 100644
--- a/tools/perf/Makefile
+++ b/tools/perf/Makefile
@@ -638,6 +638,16 @@ ifndef NO_GTK2
 	endif
 endif
 
+ifndef NO_LIBPFM
+	FLAGS_LIBPFM=$(ALL_CFLAGS) $(ALL_LDFLAGS) $(EXTLIBS) -lpfm
+	ifneq ($(call try-cc,$(SOURCE_LIBPFM),$(FLAGS_LIBPFM),libpfm4),y)
+		msg := $(warning libpfm4 not found, disables libpfm4 support. Please install libpfm4-dev);
+	else
+		BASIC_CFLAGS += -DUSE_LIBPFM
+		EXTLIBS += -lpfm
+	endif
+endif
+
 ifdef NO_LIBPERL
 	BASIC_CFLAGS += -DNO_LIBPERL
 else
diff --git a/tools/perf/builtin-list.c b/tools/perf/builtin-list.c
index 1948ece..60ca306 100644
--- a/tools/perf/builtin-list.c
+++ b/tools/perf/builtin-list.c
@@ -14,10 +14,22 @@
 #include "util/parse-events.h"
 #include "util/cache.h"
 
+#ifdef USE_LIBPFM
+#include <perfmon/pfmlib.h>
+#endif
+
 int cmd_list(int argc, const char **argv, const char *prefix __maybe_unused)
 {
 	setup_pager();
 
+#ifdef USE_LIBPFM
+	{ int ret;
+	  ret = pfm_initialize();
+	  if (ret != PFM_SUCCESS)
+		fprintf(stderr, "warning libpfm failed to initialize: %s\n", pfm_strerror(ret));
+	}
+#endif
+
 	if (argc == 1)
 		print_events(NULL, false);
 	else {
diff --git a/tools/perf/builtin-record.c b/tools/perf/builtin-record.c
index e9231659..a50a521 100644
--- a/tools/perf/builtin-record.c
+++ b/tools/perf/builtin-record.c
@@ -31,6 +31,10 @@
 #include <sched.h>
 #include <sys/mman.h>
 
+#ifdef USE_LIBPFM
+#include <perfmon/pfmlib.h>
+#endif
+
 enum write_mode_t {
 	WRITE_FORCE,
 	WRITE_APPEND
@@ -1003,6 +1007,11 @@ const struct option record_options[] = {
 	OPT_CALLBACK('j', "branch-filter", &record.opts.branch_stack,
 		     "branch filter mask", "branch stack filter modes",
 		     parse_branch_stack),
+#ifdef USE_LIBPFM
+	OPT_CALLBACK(0, "pfm-events", &record.evlist, "event",
+		     "libpfm4 event selector. use 'perf list' to list available events",
+		     parse_libpfm_events_option),
+#endif
 	OPT_END()
 };
 
@@ -1014,6 +1023,14 @@ int cmd_record(int argc, const char **argv, const char *prefix __maybe_unused)
 	struct perf_record *rec = &record;
 	char errbuf[BUFSIZ];
 
+#ifdef USE_LIBPFM
+	{ int ret;
+	  ret = pfm_initialize();
+	  if (ret != PFM_SUCCESS)
+		ui__warning("warning libpfm failed to initialize: %s\n", pfm_strerror(ret));
+	}
+#endif
+
 	evsel_list = perf_evlist__new(NULL, NULL);
 	if (evsel_list == NULL)
 		return -ENOMEM;
diff --git a/tools/perf/builtin-stat.c b/tools/perf/builtin-stat.c
index 93b9011..6712c2c 100644
--- a/tools/perf/builtin-stat.c
+++ b/tools/perf/builtin-stat.c
@@ -60,6 +60,10 @@
 #include <sys/prctl.h>
 #include <locale.h>
 
+#ifdef USE_LIBPFM
+#include <perfmon/pfmlib.h>
+#endif
+
 #define DEFAULT_SEPARATOR	" "
 #define CNTR_NOT_SUPPORTED	"<not supported>"
 #define CNTR_NOT_COUNTED	"<not counted>"
@@ -1114,6 +1118,11 @@ int cmd_stat(int argc, const char **argv, const char *prefix __maybe_unused)
 	OPT_BOOLEAN(0, "append", &append_file, "append to the output file"),
 	OPT_INTEGER(0, "log-fd", &output_fd,
 		    "log output to fd, instead of stderr"),
+#ifdef USE_LIBPFM
+	OPT_CALLBACK(0, "pfm-events", &evsel_list, "event",
+		     "libpfm4 event selector. use 'perf list' to list available events",
+		     parse_libpfm_events_option),
+#endif
 	OPT_END()
 	};
 	const char * const stat_usage[] = {
@@ -1124,6 +1133,14 @@ int cmd_stat(int argc, const char **argv, const char *prefix __maybe_unused)
 	int status = -ENOMEM, run_idx;
 	const char *mode;
 
+#ifdef USE_LIBPFM
+	{ int ret;
+	  ret = pfm_initialize();
+	  if (ret != PFM_SUCCESS)
+		fprintf(stderr, "warning libpfm failed to initialize: %s\n", pfm_strerror(ret));
+	}
+#endif
+
 	setlocale(LC_ALL, "");
 
 	evsel_list = perf_evlist__new(NULL, NULL);
diff --git a/tools/perf/config/feature-tests.mak b/tools/perf/config/feature-tests.mak
index 4add41b..d8e1b67 100644
--- a/tools/perf/config/feature-tests.mak
+++ b/tools/perf/config/feature-tests.mak
@@ -93,6 +93,19 @@ int main(void)
 endef
 endif
 
+ifndef NO_LIBPFM
+define SOURCE_LIBPFM
+#include <sys/types.h>
+#include <perfmon/pfmlib.h>
+
+int main(void)
+{
+	(void)pfm_initialize();
+	return 0;
+}
+endef
+endif
+
 ifndef NO_LIBPERL
 define SOURCE_PERL_EMBED
 #include <EXTERN.h>
@@ -203,4 +216,4 @@ int main(void)
 	return audit_open();
 }
 endef
-endif
\ No newline at end of file
+endif
diff --git a/tools/perf/util/parse-events.c b/tools/perf/util/parse-events.c
index 6b6d03e..d53db9b 100644
--- a/tools/perf/util/parse-events.c
+++ b/tools/perf/util/parse-events.c
@@ -16,6 +16,11 @@
 #include "parse-events-flex.h"
 #include "pmu.h"
 
+#ifdef USE_LIBPFM
+#include <perfmon/pfmlib_perf_event.h>
+static void print_libpfm_events(void);
+#endif
+
 #define MAX_NAME_LEN 100
 
 struct event_symbol {
@@ -1091,6 +1096,10 @@ void print_events(const char *event_glob, bool name_only)
 	}
 
 	print_tracepoint_events(NULL, NULL, name_only);
+
+#ifdef USE_LIBPFM
+	print_libpfm_events();
+#endif
 }
 
 int parse_events__is_hardcoded_term(struct parse_events__term *term)
@@ -1158,3 +1167,171 @@ void parse_events__free_terms(struct list_head *terms)
 
 	free(terms);
 }
+
+#ifdef USE_LIBPFM
+static int
+parse_libpfm_event(const char **strp, struct perf_event_attr *attr)
+{
+	const char *p;
+	size_t l;
+	int ret;
+
+	ret = pfm_get_perf_event_encoding(*strp, PFM_PLM0|PFM_PLM3,
+					  attr, NULL, NULL);
+	if (ret == PFM_SUCCESS) {
+		p = strchr(*strp, ',');
+		if (p)
+			l = p - *strp;
+		else
+			l = strlen(*strp);
+
+		*strp += l;
+	}
+	return ret;
+}
+
+int parse_libpfm_events_option(const struct option *opt, const char *str,
+			       int unset __maybe_unused)
+{
+	struct perf_evlist *evlist = *(struct perf_evlist **)opt->value;
+	struct perf_event_attr attr;
+	struct perf_evsel *evsel;
+	const char *ostr;
+	int ret;
+
+	for (;;) {
+		ostr = str;
+
+		memset(&attr, 0, sizeof(attr));
+
+		event_attr_init(&attr);
+
+		ret = parse_libpfm_event(&str, &attr);
+		if (ret != PFM_SUCCESS) {
+			fprintf(stderr, "failed to parse event %s : %s\n", str, pfm_strerror(ret));
+			return -1;
+		}
+
+		if (!(*str == 0 || *str == ',' || isspace(*str)))
+			return -1;
+
+		evsel = perf_evsel__new(&attr, evlist->nr_entries);
+		if (evsel == NULL)
+			return -1;
+
+		evsel->name = calloc(str - ostr + 1, 1);
+		if (!evsel->name) {
+			perf_evsel__delete(evsel);
+			return -1;
+		}
+		strncpy(evsel->name, ostr, str - ostr);
+
+		perf_evlist__add(evlist, evsel);
+
+		if (*str == 0)
+			break;
+		if (*str == ',')
+			++str;
+	}
+	return 0;
+}
+
+
+
+static const char *srcs[PFM_ATTR_CTRL_MAX]={
+	[PFM_ATTR_CTRL_UNKNOWN] = "???",
+	[PFM_ATTR_CTRL_PMU] = "PMU",
+	[PFM_ATTR_CTRL_PERF_EVENT] = "perf_event",
+};
+
+static void
+print_attr_flags(pfm_event_attr_info_t *info)
+{
+	int n = 0;
+
+	if (info->is_dfl) {
+		printf("[default] ");
+		n++;
+	}
+
+	if (info->is_precise) {
+		printf("[precise] ");
+		n++;
+	}
+
+	if (!n)
+		printf("- ");
+}
+
+static void
+print_libpfm_events(void)
+{
+	pfm_event_info_t info;
+	pfm_pmu_info_t pinfo;
+	pfm_event_attr_info_t ainfo;
+	const char *src;
+	int i, j, p, ret;
+
+	/* initialize to zero to indicate ABI version */
+	info.size  = sizeof(info);
+	pinfo.size = sizeof(pinfo);
+	ainfo.size = sizeof(ainfo);
+
+	putchar('\n');
+
+	pfm_for_all_pmus(p) {
+		ret = pfm_get_pmu_info(p, &pinfo);
+		if (ret != PFM_SUCCESS)
+			continue;
+
+		/* ony print events that are supported by host HW */
+		if (!pinfo.is_present)
+			continue;
+
+		/* handled by perf directly */
+		if (pinfo.pmu == PFM_PMU_PERF_EVENT)
+			continue;
+
+		for (i = pinfo.first_event; i != -1; i = pfm_get_event_next(i)) {
+
+			ret = pfm_get_event_info(i, PFM_OS_PERF_EVENT_EXT, &info);
+			if (ret != PFM_SUCCESS)
+				continue;
+
+			printf("\nName  : %s\n", info.name);
+			printf("PMU   : %s\n", pinfo.name);
+			printf("Desc  : %s\n", info.desc);
+			printf("Equiv : %s\n", info.equiv ? info.equiv : "None");
+			printf("Code  : 0x%"PRIx64"\n", info.code);
+			pfm_for_each_event_attr(j, &info) {
+				ret = pfm_get_event_attr_info(info.idx, j, PFM_OS_PERF_EVENT_EXT, &ainfo);
+				if (ret != PFM_SUCCESS)
+					continue;
+
+				if (ainfo.ctrl >= PFM_ATTR_CTRL_MAX)
+					ainfo.ctrl = PFM_ATTR_CTRL_UNKNOWN;
+
+				src = srcs[ainfo.ctrl];
+				switch(ainfo.type) {
+				case PFM_ATTR_UMASK:
+					printf("Umask : 0x%02"PRIx64" : %s: [%s] : ", ainfo.code, src, ainfo.name);
+					print_attr_flags(&ainfo);
+					printf(": %s\n", ainfo.desc);
+					break;
+				case PFM_ATTR_MOD_BOOL:
+					printf("Modif : %s: [%s] : %s (boolean)\n", src, ainfo.name, ainfo.desc);
+					break;
+				case PFM_ATTR_MOD_INTEGER:
+					printf("Modif : %s: [%s] : %s (integer)\n", src, ainfo.name, ainfo.desc);
+					break;
+				case PFM_ATTR_NONE:
+				case PFM_ATTR_RAW_UMASK:
+				case PFM_ATTR_MAX:
+				default:
+					printf("Attr  : %s: [%s] : %s\n", src, ainfo.name, ainfo.desc);
+				}
+			}
+		}
+	}
+}
+#endif
diff --git a/tools/perf/util/parse-events.h b/tools/perf/util/parse-events.h
index 2820c40..df921c1 100644
--- a/tools/perf/util/parse-events.h
+++ b/tools/perf/util/parse-events.h
@@ -34,6 +34,11 @@ extern int parse_events(struct perf_evlist *evlist, const char *str,
 extern int parse_events_terms(struct list_head *terms, const char *str);
 extern int parse_filter(const struct option *opt, const char *str, int unset);
 
+#ifdef USE_LIBPFM
+extern int parse_libpfm_events_option(const struct option *opt, const char *str,
+			       int unset);
+#endif
+
 #define EVENTS_HELP_MAX (128*1024)
 
 enum {
-- 
1.7.9.5

